WebDriverAgentLib/Utilities/FBClassChainQueryParser.m

```diff
@@ -0,0 +1,373 @@
+    return nil;
+  }
+  NSMutableArray *result = [NSMutableArray array];
+  FBBaseClassChainToken *nextToken = token;
+  for (NSUInteger charIdx = 1; charIdx < queryStringLength; ++charIdx) {
+    nextToken = [token nextTokenWithCharacter:[classChainQuery characterAtIndex:charIdx]];
+    if (nil == nextToken) {
+      [self.class tokenizationErrorWithIndex:charIdx originalQuery:classChainQuery error:error];
+      return nil;
+    }
+    if (nextToken != token) {
+      [result addObject:token];
+      token = nextToken;
+    }
+  }
+  if (nil != nextToken) {
+    [result addObject:nextToken];
+  }
+  
+  for (NSUInteger tokenIdx = 0; tokenIdx < result.count; ++tokenIdx) {
+    if ([[result objectAtIndex:tokenIdx] isKindOfClass:FBStarToken.class]) {
+      [result replaceObjectAtIndex:tokenIdx withObject:[[FBClassNameToken alloc] initWithStringValue:[FBElementTypeTransformer stringWithElementType:XCUIElementTypeAny]]];
+    }
+  }
+  
+  FBBaseClassChainToken *lastToken = [result lastObject];
+  if (!([lastToken isKindOfClass:FBClosingBracketToken.class] || [lastToken isKindOfClass:FBClassNameToken.class])) {
+    [self.class tokenizationErrorWithIndex:queryStringLength - 1 originalQuery:classChainQuery error:error];
+    return nil;
+  }
+  
+  return result.copy;
+}
+
++ (BOOL)compilationErrorWithQuery:(NSString *)originalQuery description:(NSString *)description error:(NSError **)error
+{
+  NSString *fullDescription = [NSString stringWithFormat:@"Cannot parse class chain query '%@'. %@", originalQuery, description];
+  *error = [[FBErrorBuilder.builder withDescription:fullDescription] build];
+  return YES;
+}
+
++ (nullable FBClassChain)compiledQueryWithTokenizedQuery:(NSArray<FBBaseClassChainToken *> *)tokenizedQuery originalQuery:(NSString *)originalQuery error:(NSError **)error
+{
+  NSMutableArray *result = [NSMutableArray array];
+  XCUIElementType chainElementType;
+  int chainElementPosition = 1;
+  BOOL isPositionSet = NO;
+  for (FBBaseClassChainToken *token in tokenizedQuery) {
+    if ([token isKindOfClass:FBClassNameToken.class]) {
+      @try {
+        chainElementType = [FBElementTypeTransformer elementTypeWithTypeName:token.asString];
+      } @catch (NSException *e) {
+        if ([e.name isEqualToString:FBInvalidArgumentException]) {
+          NSString *description = [NSString stringWithFormat:@"'%@' class name is unknown to WDA", token.asString];
+          [self.class compilationErrorWithQuery:originalQuery description:description error:error];
+          return nil;
+        }
+        @throw e;
+      }
+    } else if ([token isKindOfClass:FBNumberToken.class]) {
+      if (isPositionSet) {
+        NSString *description = [NSString stringWithFormat:@"Position value '%@' is expected to be set only once.", token.asString];
+        [self.class compilationErrorWithQuery:originalQuery description:description error:error];
+        return nil;
+      }
+      NSNumberFormatter *f = [[NSNumberFormatter alloc] init];
```

Comment:
Formatters are expensive to create you should do this in constructor and reuse it.